tracer/src/Datadog.Trace/Telemetry/Collectors/ProductsTelemetryCollector.cs

```diff
@@ -43,29 +45,73 @@ public void ProductChanged(TelemetryProductType product, bool enabled, ErrorData
     /// <summary>
     /// Gets all the product data recorded so far
     /// </summary>
-    public ProductsData? GetFullData() => BuildProductsData();
+    public ProductsData? GetFullData()
+    {
+        lock (_allTime)
+        {
+            var appsec = GetLatestData(_allTime, _productsByType, TelemetryProductType.AppSec);
+            var profiler = GetLatestData(_allTime, _productsByType, TelemetryProductType.Profiler);
+            var dynamicInstrumentation = GetLatestData(_allTime, _productsByType, TelemetryProductType.DynamicInstrumentation);
+
+            if (appsec is not null
+             || profiler is not null
+             || dynamicInstrumentation is not null)
+            {
+                return new ProductsData
+                {
+                    Appsec = appsec,
+                    Profiler = profiler,
+                    DynamicInstrumentation = dynamicInstrumentation,
+                };
+            }
+
+            return null;
+        }
+
+        static ProductData? GetLatestData(ProductDetail?[] allTime, ProductDetail?[] current, TelemetryProductType product)
```

Comment:
Is stylecop broken? Normally it would complain about the lack of visibility modifier

Reply:
it's a local function, no visibility modifier allowed 😄